codex/blockexchange/protobuf/message.nim

```diff
@@ -0,0 +1,194 @@
+    sendDontHave*: bool     # Note: defaults to false
+
+  Wantlist* = object
+    entries*: seq[Entry]    # A list of wantlist entries
+    full*: bool             # Whether this is the full wantlist. default to false
+
+  Block* = object
+    prefix*: seq[byte]      # CID prefix (cid version, multicodec and multihash prefix (type + length)
+    data*: seq[byte]
+
+  BlockPresenceType* = enum
+    presenceHave = 0,
+    presenceDontHave = 1
+
+  BlockPresence* = object
+    cid*: seq[byte]         # The block cid
+    `type`*: BlockPresenceType
+    price*: seq[byte]       # Amount of assets to pay for the block (UInt256)
+
+  AccountMessage* = object
+    address*: seq[byte]     # Ethereum address to which payments should be made
+
+  StateChannelUpdate* = object
+    update*: seq[byte]      # Signed Nitro state, serialized as JSON
+
+  Message* = object
+    wantlist*: Wantlist
+    payload*: seq[Block]
+    blockPresences*: seq[BlockPresence]
+    pendingBytes*: uint
+    account*: AccountMessage
+    payment*: StateChannelUpdate
+
+#
+# Encoding Message into seq[byte] in Protobuf format
+#
+
+proc write*(pb: var ProtoBuffer, field: int, value: Entry) =
+  var ipb = initProtoBuffer()
+  ipb.write(1, value.`block`)
+  ipb.write(2, value.priority.uint64)
+  ipb.write(3, value.cancel.uint)
+  ipb.write(4, value.wantType.uint)
+  ipb.write(5, value.sendDontHave.uint)
+  ipb.finish()
+  pb.write(field, ipb)
+
+proc write*(pb: var ProtoBuffer, field: int, value: Wantlist) =
+  var ipb = initProtoBuffer()
+  for v in value.entries:
+    ipb.write(1, v)
+  ipb.write(2, value.full.uint)
+  ipb.finish()
+  pb.write(field, ipb)
+
+proc write*(pb: var ProtoBuffer, field: int, value: Block) =
+  var ipb = initProtoBuffer()
+  ipb.write(1, value.prefix)
+  ipb.write(2, value.data)
+  ipb.finish()
+  pb.write(field, ipb)
+
+proc write*(pb: var ProtoBuffer, field: int, value: BlockPresence) =
+  var ipb = initProtoBuffer()
+  ipb.write(1, value.cid)
+  ipb.write(2, value.`type`.uint)
+  ipb.write(3, value.price)
+  ipb.finish()
+  pb.write(field, ipb)
+
+proc write*(pb: var ProtoBuffer, field: int, value: AccountMessage) =
+  var ipb = initProtoBuffer()
+  ipb.write(1, value.address)
+  ipb.finish()
+  pb.write(field, ipb)
+
+proc write*(pb: var ProtoBuffer, field: int, value: StateChannelUpdate) =
+  var ipb = initProtoBuffer()
+  ipb.write(1, value.update)
+  ipb.finish()
+  pb.write(field, ipb)
+
+proc ProtobufEncode*(value: Message): seq[byte] =
+  var ipb = initProtoBuffer()
+  ipb.write(1, value.wantlist)
+  for v in value.payload:
+    ipb.write(3, v)
+  for v in value.blockPresences:
+    ipb.write(4, v)
+  ipb.write(5, value.pendingBytes)
+  ipb.write(6, value.account)
+  ipb.write(7, value.payment)
+  ipb.finish()
+  ipb.buffer
+
+
+#
+# Decoding Message from seq[byte] in Protobuf format
+#
+
+proc decode*(_: type Entry, pb: ProtoBuffer): ProtoResult[Entry] =
+  var
+    value = Entry()
+    field: uint64
+  discard ? pb.getField(1, value.`block`)
```

Comment:
Why are we discarding the result here?

Reply:
let's check an example you provided: https://github.com/status-im/nim-libp2p/blob/unstable/libp2p/protocols/pubsub/rpc/protobuf.nim#L243

and the definition: https://github.com/status-im/nim-libp2p/blob/unstable/libp2p/protobuf/minprotobuf.nim#L444

IF getField finds field with given number in the buffer, it assigns decoded value to the variable, otherwise it is kept unchanged. And then it return ?!bool to signal a) decoding error , b) whether it found&assigned this field